src/decode.jl

```diff
@@ -0,0 +1,147 @@
+        LibJpeg.jpeg_read_header(cinfo_ptr, true)
+
+        # set decompression parameters, if given
+        r = _cal_scale_ratio(scale_ratio)
+        cinfo.scale_num, cinfo.scale_denom = r.num, r.den
+        cinfo.out_color_space = jpeg_cls
+
+        # eagerly calculate dimension information so that `output_XXX` fields are valid.
+        LibJpeg.jpeg_calc_output_dimensions(cinfo_ptr)
+        out_size = (Int(cinfo.output_width), Int(cinfo.output_height))
+        if !all(x -> x<=65535, out_size)
+            error("Suspicious inferred image size $out_size: each dimension is expected to have at most 65535 pixels.")
+        end
+        out_ndims = Int(cinfo.output_components)
+        @assert out_ndims == length(out_CT) "Suspicous output color space: $cinfo.out_color_space"
+
+        out = Matrix{out_CT}(undef, out_size)
+        _jpeg_decode!(out, cinfo)
+    finally
+        LibJpeg.jpeg_destroy_decompress(Ref(cinfo))
+        ccall(:fclose, Cint, (Ptr{Libc.FILE},), infile)
+    end
+
+    if out_CT <: CT
+        return transpose ? out : permutedims(out, (2, 1))
+    else
+        return transpose ? CT.(out) : CT.(PermutedDimsArray(out, (2, 1)))
+    end
+end
+function jpeg_decode(filename::AbstractString; kwargs...)
+    return jpeg_decode(_default_out_color_space(filename), filename; kwargs...)
+end
+
+function _jpeg_decode!(out::Matrix{<:Colorant}, cinfo::LibJpeg.jpeg_decompress_struct)
+    cinfo_ptr = Ref(cinfo)
+
+    row_stride = size(out, 1) * length(eltype(out))
+    buf = Vector{UInt8}(undef, row_stride)
+    out_uint8 = reinterpret(UInt8, out)
+
+    LibJpeg.jpeg_start_decompress(cinfo_ptr)
+    while cinfo.output_scanline < cinfo.output_height
+        # TODO(johnnychen94): try if we can directly write to `out` without using `buf`
+        GC.@preserve buf LibJpeg.jpeg_read_scanlines(cinfo_ptr, Ref(pointer(buf)), 1)
+        copyto!(out_uint8, (cinfo.output_scanline-1) * row_stride + 1, buf, 1, row_stride)
+    end
+    LibJpeg.jpeg_finish_decompress(cinfo_ptr)
+
+    return out
+end
+
+# libjpeg-turbo only supports ratio M/8 with M from 1 to 16
+const _allowed_scale_ratios = ntuple(i->i//8, 16)
+_cal_scale_ratio(r::Real) = _allowed_scale_ratios[findmin(x->abs(x-r), _allowed_scale_ratios)[2]]
+
+function _default_out_color_space(filename::AbstractString)
+    infile = ccall(:fopen, Libc.FILE, (Cstring, Cstring), filename, "rb")
+    @assert infile.ptr != C_NULL
+    local cinfo
+    try
+        cinfo = LibJpeg.jpeg_decompress_struct()
+        cinfo_ptr = Ref(cinfo)
+        jerr = Ref{LibJpeg.jpeg_error_mgr}()
+        cinfo.err = LibJpeg.jpeg_std_error(jerr)
+        LibJpeg.jpeg_create_decompress(cinfo_ptr)
+        LibJpeg.jpeg_stdio_src(cinfo_ptr, infile)
+        LibJpeg.jpeg_read_header(cinfo_ptr, true)
+        LibJpeg.jpeg_calc_output_dimensions(cinfo_ptr)
+        return jpeg_color_space(cinfo.out_color_space)
+    finally
+        LibJpeg.jpeg_destroy_decompress(Ref(cinfo))
```

Comment:
```suggestion
        LibJpeg.jpeg_destroy_decompress(cinfo_ptr)
```
For this specific case, it doesn't matter. But in general, you should not simply translate
```C
jpeg_destroy_decompress(&cinfo)
```
to
```julia
jpeg_destroy_decompress(Ref(cinfo))
```

A more appropriate translation is  
```Julia
cinfo_ref = Ref(cinfo)
ret = jpeg_destroy_decompress(cinfo_ref)
cinfo = cinfo_ref[]
ret
```

Reply:
> For this specific case, it doesn't matter.

Just a confirmation, it doesn't matter because this is a destroy/deconstruct step and `cinfo` is no longer used after this?

Reply:
Right.